lib/salus/scanners/osv/base.rb

```diff
@@ -0,0 +1,152 @@
+require 'salus/scanners/base'
+require 'json'
+require 'zip'
+require 'uri'
+require 'net/http'
+
+module Salus::Scanners::OSV
+  class Base < Salus::Scanners::Base
+    def run
+      raise NoMethodError, 'implement in subclass'
+    end
+
+    def should_run?
+      raise NoMethodError, 'implement in subclass'
+    end
+
+    def name
+      self.class.name.sub('Salus::Scanners::OSV::', '')
+    end
+
+    private
+
+    def osv_vulnerabilities
+      @osv_vulnerabilities ||= fetch_vulnerabilities
+    end
+
+    def osv_url
+      urls = []
+      # Bucket contains individual entries for an ecosystem in OSVF.
+      urls.append(osv_url_for("Go")) if @repository.go_sum_present? || @repository.go_mod_present?
+      urls.append(osv_url_for("PyPI")) if @repository.requirements_txt_present?
+      urls.append(osv_url_for("Maven")) if @repository.pom_xml_present?
+
+      urls
+    end
+
+    def osv_url_for(package)
+      # Approximate zip sizes
+      # Go: all.zip (478 KB) -> 1.1 MB
+      # PyPI: all.zip (3.4 MB) -> 10.9 MB
+      # Maven: all.zip (1.6 MB) -> 5.1 MB
+      url = "https://osv-vulnerabilities.storage.googleapis.com/"\
+      + package + "/all.zip"
+      URI(url)
+    end
+
+    def fetch_vulnerabilities
+      results = []
+      all_vulnerabilities_found = []
+
+      # Flatten vulnerabilities by package name.
+      all_vulnerabilities = send_request
+
+      all_vulnerabilities.each do |vulnerability|
+        all_vulnerabilities_found.append(flatten_by_affected(vulnerability))
+      end
+      all_vulnerabilities_found.flatten!
+
+      # Handle removing exception ids from vulnerabilities found.
+      exception_ids = fetch_exception_ids
+      if exception_ids.any? && all_vulnerabilities_found.any?
+        all_vulnerabilities_found.delete_if do |v|
+          identifiers_found = v.fetch("aliases", []) + [v.fetch("id")]
+          intersection = identifiers_found & exception_ids
+          true if intersection.length.positive?
+        end
+      end
+
+      # Group vulnerabilities by aliases / id
+      # Picks 1 item from each grouped item
+      # { "ID-1": [
+      #      {"name": "sample-dep-1"},  {"name": "sample-dep-1"}
+      #    ]
+      # }
+      grouped = all_vulnerabilities_found.group_by { |d| d.fetch("aliases", [d.fetch("id")]) }
```

Comment:
```
results = []
grouped = all_vulnerabilities_found.group_by { |d| d.fetch("aliases", [d.fetch("id")]) }
```
Please declare `results` close to it's first usage.

Reply:
Gonna update this wherever I declare variables at the top.